### PR TITLE
Hyper Func

### PR DESCRIPTION
**Pull request by imeoer**
part of the work, don't merge yet.

### CODE REVIEW

**Comment by Gnep** [comment]
@imeoer 

- seems missing `timeout` option in `hyper func create` and `hyper func update`
-

**Comment by Gnep** [comment]
@imeoer 

Plz ask @jmknyc06 or @mrmrcoleman to review.

**Comment by imeoer** [comment]
@Gnep Will do that.

File: FAQ/quota_and_limits.md [hunk 1]
```diff
@@ -21,8 +25,9 @@ Hyper.sh has certain hard and soft (`quota`) limits in using its service. Hard l
 	- Snapshot: 40 snapshots per region
 - Floating IP: 2 per region
 - Credential: 3 per region
```
Comment (mrmrcoleman) [inline FAQ/quota_and_limits.md]: This should be 'Credentials'

File: Feature/container/func.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+Hyper Func is a Docker-centric Serverless platform. You can wrap functions in Docker images and have them run on demand.
+
+- Fire and forget
+	- Fully managed service
+    - Easy to use for developers
+    - Minimal ops overhead
+- Docker centric
+	- Deploy your function in Docker image
+	- Any language, any libs, any deps
```
Comment (mrmrcoleman) [inline Feature/container/func.md]: Any language, any libraries, any dependencies

File: Feature/container/func.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+- Secure Container Runtime
+	- VM-like isolation
+	- New container instance for each function call
+	- Run as long as you wish
```
Comment (mrmrcoleman) [inline Feature/container/func.md]: 'Run functions for as long as you wish'

File: Feature/container/func.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+## How it works
+
+1. Hyper Func uses the popular Docker images as the format to deploy functions. Baking the code, dependencies, data into Docker images, and then register the images to create functions.
```
Comment (mrmrcoleman) [inline Feature/container/func.md]: Hyper Func uses straight Docker images as the underlying format to deploy functions. Code, dependencies and data are baked into Docker images and then uploaded to any popular container registry. From there you can launch functions with the runtime of your choosing.

File: Feature/container/func.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+2. Upon calling, a new container will be launched from the function image. The HTTP request payload is passed to the container STDIN as the function input. A new call ID will be returned, which can be used to retrieve the output later.
```
Comment (mrmrcoleman) [inline Feature/container/func.md]: Upon launch a new container is started with the selected function image. The HTTP request payload is passed to the container STDIN as input to the function and a call ID is returned so you can retrieve output later.

File: Feature/container/func.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+3. The max concurrent function calls a user could execute at a given time is subject to the user's quota. When the max concurrency is reached, new calls will be queued to wait for slots.
```
Comment (mrmrcoleman) [inline Feature/container/func.md]: The maximum number of concurrent functions is subject to the user's quota, which can be upgraded on request. When the quota limit is reached subsequent functions are queued until a slot becomes available.

File: Feature/container/func.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+4. The queued calls are processed in the ***First-In-First-Out*** manner. However we cannot guarantee the function execution will be finished in such order.
```
Comment (mrmrcoleman) [inline Feature/container/func.md]: Queued functions are processed in a ***First-In-First-Out*** order. We cannot however guarantee that the functions will terminate in this order however.

File: Feature/container/func.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+5. Hyper Func maintains a 50MB cache for each function (**not call**). The cache is used to store the STDOUT of **finished** function calls. These data need to be fetched by `hyper func get`, and then will be removed once be fetched. Otherwise the cache will be rotated once full.
```
Comment (mrmrcoleman) [inline Feature/container/func.md]: Hyper Func maintains a 50MB cache for each individual function. (***Not per function call***). This cache is used to store the STDOUT of **finished** function calls. This data therefore needs to be fetched regularly using `hyper func get`. Once the 50MB limit is reached the cache will be rotated.

File: Feature/container/func.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+6. Hyper Func maintains a 50MB cache for logs of each function. These data need to be fetched by `hyper func logs`. Otherwise the cache will be rotated once full.
```
Comment (mrmrcoleman) [inline Feature/container/func.md]: Hyper Func maintains a 50MB cache for the logs of each function. This data needs to be fetched using `hyper func logs`. Once the cache is full the logs will be rotated.

File: Feature/container/func.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+7. For each function call, the max data size of STDIN and STDOUT is 5MB.
```
Comment (mrmrcoleman) [inline Feature/container/func.md]: For each function call, the maximum size of data flowing through STDIN and STDOUT is limited to 5MB.

File: Feature/container/func.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+8. The headers of HTTP request and the env config of function will be as the environment variable of container.
```
Comment (mrmrcoleman) [inline Feature/container/func.md]: The headers of the HTTP request and any environment configuration of the function will be exposed as environment variables in the host container.

File: Feature/container/func.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+$ hyper func create --name helloworld ubuntu echo HelloWorld
+helloworld is created with the address of https://us-west-1.hyperfunc.io/call/helloworld/e62c014e-386c-42ea-8d07-41d44e98cc3d
+```
+
+2. Call the function:
+``` bash
+$ hyper func call helloworld
+CallId: 218b7b10-e7f1-4c48-ac3c-66792f8ffc06
+```
+> Tips: calling with payload `echo Hello | hyper func call helloworld`
```
Comment (mrmrcoleman) [inline Feature/container/func.md]: Tips calling with payload `echo Hello | hyper func call helloworld`

File: Feature/container/func.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+2017-03-16T15:06:10Z [FINISHED] CallId: 218b7b10-e7f1-4c48-ac3c-66792f8ffc06, ShortStdout: HelloWorld
+, ShortStderr: 
+```
+
+4. Retrieve the function return:
+``` bash
+$ hyper func get --wait 218b7b10-e7f1-4c48-ac3c-66792f8ffc06
+HelloWorld
+```
+> Tips: `--wait` blocks the CLI until the call finished (or failed)
```
Comment (mrmrcoleman) [inline Feature/container/func.md]: Tip: `--wait` blocks the CLI until the call finished (or failed)

File: Reference/API/2016-04-04 [Ver. 1.23]/Func/inspect.md [hunk 14]
```diff
@@ -0,0 +1,54 @@
+# Inspect a function
+
+`GET /funcs/$name`
+
+Inspect a function to get the detailed information.
```
Comment (mrmrcoleman) [inline Reference/API/2016-04-04 [Ver. 1.23]/Func/inspect.md]: Inspect a function to get more details information.

File: Reference/API/2016-04-04 [Ver. 1.23]/Func/status.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
+# Retrive status of a function
```
Comment (mrmrcoleman) [inline Reference/API/2016-04-04 [Ver. 1.23]/Func/status.md]: #Retrieve the status of a function

File: FAQ/quota_and_limits.md (hunk 1 above)
Comment (mrmrcoleman) [inline FAQ/quota_and_limits.md]: Should be 'Credentials'

File: Feature/container/func.md (hunk 8 above)
Comment (imeoer) [inline Feature/container/func.md]: @mrmrcoleman Seems lost `will be removed once be fetched`.

**Pull request merged**